core/src/avm2/activation.rs

```diff
@@ -2899,17 +2890,29 @@ impl<'a, 'gc> Activation<'a, 'gc> {
 
         let value = self.pop_stack();
 
-        if let Ok(value) = value.coerce_to_object(self) {
-            let is_instance_of = value.is_instance_of(self, type_object)?;
+        let class = type_object.as_class_object();
 
-            self.push_stack(is_instance_of);
-        } else if matches!(value, Value::Undefined) {
-            // undefined
-            return Err(make_null_or_undefined_error(self, value, None));
-        } else {
-            // null
-            self.push_stack(false);
-        }
+        let res = match value {
+            Value::Null => false,
+            Value::Undefined => {
+                return Err(make_null_or_undefined_error(self, value, None));
+            }
+            Value::Object(obj) => obj.is_instance_of(self, type_object)?,
+            Value::Number(_) => {
+                class == Some(self.avm2().classes().number)
+                    || class == Some(self.avm2().classes().object)
+            }
+            Value::Bool(_) => {
+                class == Some(self.avm2().classes().boolean)
+                    || class == Some(self.avm2().classes().object)
+            }
+            Value::String(_) => {
+                class == Some(self.avm2().classes().string)
+                    || class == Some(self.avm2().classes().object)
+            }
+            Value::Integer(_) => panic!("How do we do instanceof for integers?"),
```

Comment:
I'm pretty sure this is `class == Some(self.avm2().classes().object)`- the other number/int/uint classes return `false`.

Reply:
For future reference, this was wrong- it returns true for both `Number` and `Object`, just like `Value::Number`